### ISSUE TITLE
rootcling/root-io error loading std::filesystem::path

### ISSUE THREAD

**Issue by giraudpf** [root-project/root#13574]
### Check duplicate issues.

- [X] Checked for duplicates

### Description

Rootcling fails to load correctly std::filesystem::path.

This is discussed in [this post](https://root-forum.cern.ch/t/rootcling-with-std-filesystem/55388), from which I am reproducing here the provided example.

### Reproducer

Create the following two files:

test.hh:
```C++
#ifndef test_hh
#define test_hh
#include <iostream>
#include <filesystem>
namespace fs = std::filesystem;

class test
{
public:
    test();
    ~test();

private:
    fs::path path;
};

#endif
```

test.cpp:
```C++
#include "test.hh"
test::test()
{
    std::cout << "test constructor" << std::endl;
    this->path = ".";
}

test::~test()
{
    std::cout << "test destructor" << std::endl;
}
```

Then load with ACLiC:
```
$ root
   ------------------------------------------------------------------
  | Welcome to ROOT 6.28/06                        https://root.cern |
  | (c) 1995-2023, The ROOT Team; conception: R. Brun, F. Rademakers |
  | Built for linuxx8664gcc on Aug 30 2023, 16:45:00                 |
  | From tags/v6-28-06@v6-28-06                                      |
  | With c++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0                   |
  | Try '.help'/'.?', '.demo', '.license', '.credits', '.quit'/'.q'  |
   ------------------------------------------------------------------

root [0] .L test.cpp+
Info in <TUnixSystem::ACLiC>: creating shared library /data/giraudpf/tmp/./test_cpp.so
input_line_11:6:22: error: '_List' is a private member of 'std::filesystem::path'
   filesystem::path::_List::_Impl_deleter _1;
                     ^
/usr/include/c++/11/bits/fs_path.h:636:12: note: declared private here
    struct _List
           ^
input_line_11:7:22: error: '_List' is a private member of 'std::filesystem::path'
   filesystem::path::_List::_Impl* _0;
                     ^
/usr/include/c++/11/bits/fs_path.h:636:12: note: declared private here
    struct _List
           ^
Error in <Load>: Could not declare TEmulatedTuple<filesystem::path::_List::_Impl*,filesystem::path::_List::_Impl_deleter>
```

### ROOT version

```
   ------------------------------------------------------------------
  | Welcome to ROOT 6.28/06                        https://root.cern |
  | (c) 1995-2023, The ROOT Team; conception: R. Brun, F. Rademakers |
  | Built for linuxx8664gcc on Aug 30 2023, 16:45:00                 |
  | From tags/v6-28-06@v6-28-06                                      |
  | With c++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0                   |
  | Try '.help'/'.?', '.demo', '.license', '.credits', '.quit'/'.q'  |
   ------------------------------------------------------------------

```

### Installation method

Built from source

### Operating system

Ubuntu 22.04

**Comment by hahnjo** [comment]
Hi @giraudpf, sorry for the long silence on this. I can reproduce the problem and can confirm it happens in the latest `master` and GCC 13.2.0 as well. Interestingly, everything is fine once `test.cpp+` has been compiled, ie if you start a fresh `root` prompt and input `.L test.cpp+` again.

The error comes from
https://github.com/root-project/root/blob/0865d47e019470bc714d8166a376706c317de4e2/core/metacling/src/TCling.cxx#L3909
declaring
```c++
#ifndef ROOT_INTERNAL_TEmulated_TEmulatedTuplelEfilesystemcLcLpathcLcL_ListcLcL_ImplmUcOfilesystemcLcLpathcLcL_ListcLcL_Impl_deletergR                                                                 
#define ROOT_INTERNAL_TEmulated_TEmulatedTuplelEfilesystemcLcLpathcLcL_ListcLcL_ImplmUcOfilesystemcLcLpathcLcL_ListcLcL_Impl_deletergR                                                                 
namespace ROOT { namespace Internal {
template <class... Types> struct TEmulatedTuple;
template <> struct TEmulatedTuple<filesystem::path::_List::_Impl*,filesystem::path::_List::_Impl_deleter> {                                                                                            
   filesystem::path::_List::_Impl_deleter _1;
   filesystem::path::_List::_Impl* _0;
};
}}
#endif
```

Adding @pcanal as the original author of `AtlernateTuple` - I guess we shouldn't be doing this for private members, such as of `std::filesystem::path`?

**Comment by hahnjo** [comment]
Right, that's also what I was scratching my head. By putting an `assert(0)`, I found that `AlternateTuple` is called from a `rootcling` invocation, and the member that triggers the `tuple` is a `unique_ptr<filesystem::path::_List::_Impl,filesystem::path::_List::_Impl_deleter>`. ie the underlying cause is the same as https://github.com/root-project/root/issues/13361

**Comment by pcanal** [comment]
> ie the underlying cause is the same as https://github.com/root-project/root/issues/13361

Ok but I am still missing the link between that `unique_ptr` and `std::tuple` :(

**Comment by hahnjo** [comment]
> Ok but I am still missing the link between that `unique_ptr` and `std::tuple` :(

`std::tuple` is the internal data structure used in `std::unique_ptr`; see also https://github.com/root-project/root/pull/13134#issuecomment-1618637853

**Comment by giraudpf** [comment]
I confirm that this issue is fixed with ROOT release 6.32.02

However, I noticed that there is no mention of this in the release notes at https://root.cern/doc/v632/release-notes.html#release-6.32.02

Many thanks for this update!
